### ISSUE TITLE
☂️ File System Route API Umbrella Issue

### ISSUE THREAD

**Issue by blainekasten** [gatsbyjs/gatsby#26375]
People experimenting with File System Route API — this is the place to ask questions, report bugs, suggest ideas and collaborate on improvements.

**Quick links:**

- [Docs](https://www.gatsbyjs.com/docs/file-system-page-creation/)
- [Example](https://github.com/gatsbyjs/gatsby/tree/master/examples/route-api)

**Known issues/Feature requests:**

- An API to customize `pageContext` and/or pass other variables to the underlying `createPage`call
- Resolve field that holds an array of X and create pages for it
- Ability to use a custom field/variable for the file name, e.g. `{foobar}.js` and define the shape (e.g. currently `{Model.id}`) inside the file
	- This is quite important for theme shadowing

**Comment by t2ca** [comment]
I followed your example code and I'm getting the following error. However when i tried the pre release (`yarn add gatsby@unifiedroutes`) it was working perfectly fine.
```
 ERROR #11321  PLUGIN

"gatsby-plugin-page-creator" threw an error while running the onPreInit lifecycle:

CollectionBuilderError: An error occured extracting the Model from the slug parts. This is likely a bug within Gatsby and not your code. Please report it to us if you run into this.

  186 |                         absolutePath = require.resolve(_path.default.join(pagesPath, relativePath));
  187 |                         _context2.next = 3;
> 188 |                         return (0, _collectionExtractQueryString.collectionExtractQueryString)(absolutePath);
      |                                                                                               ^
  189 |
  190 |                       case 3:
  191 |                         queryString = _context2.sent;

File: node_modules/gatsby-plugin-page-creator/gatsby-node.js:188:95



  Error: CollectionBuilderError: An error occured extr  acting the Model from the slug parts. This is likely   a bug within Gatsby and not your code. Please repor  t it to us if you run into this.

  - path-utils.js:19 extractModel
    [t2.ca]/[gatsby-plugin-page-creator]/path-utils.js    :19:11

  - collection-extract-query-string.js:23 collectionEx    tractQueryString
    [t2.ca]/[gatsby-plugin-page-creator]/collection-ex    tract-query-string.js:23:47

  - gatsby-node.js:188 _callee2$
    [t2.ca]/[gatsby-plugin-page-creator]/gatsby-node.j    s:188:95

  - new Promise

  - gatsby-node.js:214
    [t2.ca]/[gatsby-plugin-page-creator]/gatsby-node.j    s:214:30

  - Array.map

  - gatsby-node.js:179 _callee3$
    [t2.ca]/[gatsby-plugin-page-creator]/gatsby-node.j    s:179:38

  - From previous event:

  - api-runner-node.js:440 Promise.catch.decorateEvent    .pluginName
    [t2.ca]/[gatsby]/src/utils/api-runner-node.js:440:    9

  - From previous event:

  - api-runner-node.js:439
    [t2.ca]/[gatsby]/src/utils/api-runner-node.js:439:    14

  - timers.js:458 processImmediate
    internal/timers.js:458:21

  - From previous event:

  - api-runner-node.js:431
    [t2.ca]/[gatsby]/src/utils/api-runner-node.js:431:    13

  - From previous event:

  - api-runner-node.js:347 module.exports
    [t2.ca]/[gatsby]/src/utils/api-runner-node.js:347:    3

  - initialize.ts:191 initialize
    [t2.ca]/[gatsby]/src/services/initialize.ts:191:9
```
also

```
 ERROR

Collection page builder encountered an error parsing the filepath. To use collection paths the schema to follow is {Model.field}. The problematic part is: {label}.js.
filePath: /Users/admin/Desktop/Projects/t2/t2.ca/src/pages/product/{label}.js

failed createPagesStatefully - 0.144s
```

**Comment by blainekasten** [comment]
It looks like the reason for the failure is that you are missing the graphql model in the filepath. You have `{label}.js` but it needs to be something like `{Model.label}.js`

**Comment by blainekasten** [comment]
@noskovvkirill I'm expecting to see some specific logs if you are using collection routes and I'm not seeing them. What is the file system path of the collection or client route you expected to see? You mentioned `/users/[id]` I'm assuming you have a file that is `src/pages/users/[id].js` ? If so, what happens when you visit `localhost:8000/users/123`

**Comment by czabaj** [comment]
Hi there,

## Found 1 bug in the documentation.

[The Routing and linking](https://www.gatsbyjs.com/docs/file-system-page-creation/#routing-and-linking) section contains an example with double-colon `products/{Product:name}`, a dot must be there (there are two occurrences of this bug).

Also, the example query could probably reflect better Relay convention -- nesting data in `{ edges { node { ... } } }`.

I'm very happy with GQL errors, which provides useful hints for `filePath` argument 👍 

## Encountered one caveat with `gatsbyPath` field

I have a problem with [gatsby-plugin-page-creator](https://www.gatsbyjs.com/plugins/gatsby-plugin-page-creator/?=page-creator). If some page needs specific components, utilities, or other files, we colocate all its auxiliaries in its directory. E.g. page `/shop` lives in `shop/index.js` and its specific React components lies in `shop/components/`. This is so called [Deletable React application design](https://medium.com/@shanplourde/deletable-react-application-design-75b3862d9759). Obviously, we want these auxiliary files to be ignored by gatsby-plugin-page-creator but current default ignore pattern are insufficient and built in configuration of gatsby-plugin-page-creator cannot be overridden, so I configured new gatsby-plugin-page-creator plugin, which sources `src/web/` instead of `src/pages/` and I configured the ignore pattern to match also all _directories_ which starts with a underscore (by default, only _files_ with leading underscore are ignored).

The `filePath` argument of `gatsbyPath` field counts with sourcing the files from `src/pages/`, you must enter absolute path, which is relative to `src/pages/`. I could hack this with absolute route containing double dot, e.g. `/../web/shop/{Product.name}`, but it is not obvious, nor readable.

I would prefer to use Gatsbys default `pages/` directory, I will look into ways how to override default ignore pattern, but sourcing different path, than `pages/` is currently hacky 😕

**Comment by noskovvkirill** [comment]
> @noskovvkirill I'm expecting to see some specific logs if you are using collection routes and I'm not seeing them. What is the file system path of the collection or client route you expected to see? You mentioned `/users/[id]` I'm assuming you have a file that is `src/pages/users/[id].js` ? If so, what happens when you visit `localhost:8000/users/123`

Excuse me, I simply didn't have the last version of Gatsby. It works now. Thank you!

**Comment by blainekasten** [comment]
Thanks for all the great feedback @czabaj!

I've fixed the documentation and i'll look into the problem you mentioned with `gatsbyPath`. Now that I'm thinking about it I definitely did not think about adding support for the `ignore` option

**Comment by czabaj** [comment]
Is it possible to use FS routes to generate collection pages from nested data?

For example, consider nested collection (simplified):

```JavaScript
Products = [
    { category: 'shirts', items: ['blue', 'white', 'yellow'] },
    { category: 'pants', items: ['blue', 'lime', 'dodgerblue'] }
]
```

I want to translate such data into routes:

```
/product/shirts
/product/pants
/product/shirts/blue
/product/shirts/white
/product/shirts/yellow
/product/pants/blue
/product/pants/lime
/product/pants/dodgerblue
```

Creating category pages is simple, just create file `/product/{Product.category}`.

But how dynamically create nested collection routes?

**Comment by czabaj** [comment]
Found another minor issue with [the sitemap plugin](https://www.gatsbyjs.com/plugins/gatsby-plugin-sitemap/) - currently, it exposes dynamic routes.

E.g. `/src/pages/users/[id].js` creates record in sitemap.xml

```
<url>
   <loc>https://example.com/users/[id]/</loc>
</url>
```

In my case, I don't want to expose dynamic pages in Sitemap at all and I think this is the common goal.

This config handles it

```
{
	resolve: `gatsby-plugin-sitemap`,
	options: {
		serialize: ({
			site: {
				siteMetadata: { siteUrl },
			},
			allSitePage,
	}) => {
			const reDynamicPage = /\[[\w.]+\]\/$/
			return allSitePage.edges
				.filter(
					(edge) =>
						// remove empty nodes
						edge &&
						// filter out all dynamic pages created with FS API
						!reDynamicPage.test(edge.node.path)
				)
				.map(({ node: { path } }) => {
					return {
						url: siteUrl + path,
						changefreq: `weekly`, // adjust to your preferences
						priority: 1, // adjust to your preferences
					};
				});
		},
	},
},
```

**Comment by czabaj** [comment]
Regarding [nested collections](https://github.com/gatsbyjs/gatsby/issues/26375?utm_campaign=Product%20Announcements&utm_medium=email&_hsmi=93142808&_hsenc=p2ANqtz-9oZ_ktXu9NqNWO0-1mFopXYWrOmrgrmtrMfiTnQUEqCsKWiqWPSGF9DTqTOqaJMc2VVU6LeVyZlu9eNFusZOTS9gRuyA&utm_content=93142808&utm_source=hs_automation#issuecomment-689531533), how about support also dynamic _directories_? Given example above, one could accomplish the desired URIs with file structure `/src/pages/product/{Product.category}/{Product.items.name}.js`, where `{Product.items.name}.js` would implicitly filter for `Product` with category of currently generated `{Product.category}`?

**Comment by andreme** [comment]
@blainekasten I'm trying to create my pages from a gatsby-source-graphql data source and have run into the problem that my schema does not have a nodes list, which is forced on me by the CollectionPagesQueryFragment.
Why can't I write my query without it?
The file name also still requires the Model name? I would have thought using unstable_collectionGraphql would remove that requirement.

**Comment by ynnoj** [comment]
I was checking out the new file system routing APIs and ran into some trouble when trying to use the generated `gatsbyPath` field on types from [`gatsby-source-graphcms`](https://github.com/GraphCMS/gatsby-source-graphcms). The source plugin uses the new `gatsby-grpahql-toolkit`.

```js
export const pageQuery = graphql`
  query IndexPageQuery {
    products: allGraphCmsProduct {
      nodes {
        name
        gatsbyPath(filePath: "./{GraphCmsProduct.slug}")
      }
    }
  }
`;
```
Unfortunately the query above errors:

`error  Cannot query field "gatsbyPath" on type "GraphCMS_Product"`

There's a full reproduction available [here](https://github.com/GraphCMS/graphcms-examples/pull/158). Thanks!

**Comment by LekoArts** [comment]
To all previous comments: I'm taking over this feature and slowly getting up to speed. Eventually I'll get back to you and look into the issue :)

---

@ynnoj Thanks for the report! Just a quick check: Does it change anything when you remove the `.` in the path?
The `gatsbyPath` field gets [added here](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-page-creator/src/gatsby-node.ts#L131-L182) so e.g. the problem could be that the type isn't there yet when it tries to add it.

**Comment by ynnoj** [comment]
> @ynnoj Thanks for the report! Just a quick check: Does it change anything when you remove the . in the path?
The gatsbyPath field gets added here so e.g. the problem could be that the type isn't there yet when it tries to add it.

Same issue unfortunately. Is the path relative or absolute?

**Comment by LekoArts** [comment]
Current [requirement](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-page-creator/src/validate-path-query.ts#L8-L12) is to start it with a `/` so that's why I asked :)

**Comment by ynnoj** [comment]
@LekoArts So the `/` (root) would be the `src/pages` directory?

**Comment by LekoArts** [comment]
Yeah, correct! Thanks for the question, I'll note this to mention it somewhere in the upcoming docs 😅

**Comment by LekoArts** [comment]
@ynnoj PR for a fix is up: https://github.com/gatsbyjs/gatsby/pull/27487 - will let you know once it's published.

**Comment by LekoArts** [comment]
@czabaj Reading your comment in https://github.com/gatsbyjs/gatsby/issues/26375#issuecomment-680075683 - am I'm reading it right that the only problem left is that `src/pages` is hardcoded in `gatsbyPath` ?

**Comment by LekoArts** [comment]
@ynnoj Please try out `gatsby@2.24.79`

**Comment by ynnoj** [comment]
@LekoArts Resolved, thank you! 🚀

**Comment by czabaj** [comment]
@LekoArts correct, fs route API shall resolve all paths configured in all instances of `gatsby-plugin-page-creator`, but this is not major issue.

**Comment by LekoArts** [comment]
> Regarding [nested collections](https://github.com/gatsbyjs/gatsby/issues/26375?utm_campaign=Product%20Announcements&utm_medium=email&_hsmi=93142808&_hsenc=p2ANqtz-9oZ_ktXu9NqNWO0-1mFopXYWrOmrgrmtrMfiTnQUEqCsKWiqWPSGF9DTqTOqaJMc2VVU6LeVyZlu9eNFusZOTS9gRuyA&utm_content=93142808&utm_source=hs_automation#issuecomment-689531533), how about support also dynamic _directories_? Given example above, one could accomplish the desired URIs with file structure `/src/pages/product/{Product.category}/{Product.items.name}.js`, where `{Product.items.name}.js` would implicitly filter for `Product` with category of currently generated `{Product.category}`?

@czabaj I've fixed a bug in https://github.com/gatsbyjs/gatsby/pull/27684 that will enable this to work. You can see an example here: https://github.com/gatsbyjs/gatsby/blob/unified-routes/improve-example/examples/route-api/src/pages/parks/%7Bpark.meta__location__type%7D/%7Bpark.name%7D.js

**Comment by KyleAMathews** [comment]
Found a bug.

Path segments don't work if they include other text.

E.g. `/src/pages/test-{MarkdownRemark.fields__slug}.js` doesn't work but `/src/pages/{MarkdownRemark.fields__slug}.js`

It doesn't generate the query with the `fields { slug }` portion — the error you see is:

```
PageCreator: Could not find value in the following node for key {MarkdownRemark.fields__slug} (transformed to fields.slug)

{
    "id": "f27e27e8-2dd6-5cde-866e-d015dc9c9dbd"
}
```

On a related note — when there's an error on every page of the collection, it'd be great to group the errors instead of printing out one for each page (110 in the case of my blog).

**Comment by KyleAMathews** [comment]
Bug report or maybe Feature request:

I have a field on nodes that's an array of strings — I expect to be able to write `MarkdownRemark.fields__tagArray` and get a unique list of tag pages — one for each tag. As it is, the generated URLs are concatenating the array items together. Which doesn't seem what anyone would want ever.

**Comment by KyleAMathews** [comment]
I got a unhelpful issue when I made a mistake in the syntax:

```
 ERROR
PageCreator: Syntax Error: Cannot parse the unexpected character ".".
```

I wrote `{MarkdownRemark.fields.slug}.js`  instead of `{MarkdownRemark.fields__slug}.js`

I'd expect the error to be something more like — "hey, `MarkdownRemark` is a collection & these are the fields on it & how you'd select them" or something along those lines.

**Comment by LekoArts** [comment]
> Found a bug.
> Path segments don't work if they include other text.

Yep, a thing that is not supported at the moment.

> On a related note — when there's an error on every page of the collection, it'd be great to group the errors instead of printing out one for each page (110 in the case of my blog).

![image](https://user-images.githubusercontent.com/16143594/97592537-0d939f80-1a01-11eb-943d-8b9be7cfaa58.png)

Grouped it under one message, and with `--verbose` the previous output (for every page) will be output.

Will group other possible errors, too.

> I'd expect the error to be something more like — "hey, MarkdownRemark is a collection & these are the fields on it & how you'd select them" or something along those lines.

👍

**Comment by t2ca** [comment]
hey there,
Is it possible to insert props into pageContext like we did in gatsby-node with createPage?

**Comment by LekoArts** [comment]
The new documentation is merged into the main PR branch now: https://github.com/gatsbyjs/gatsby/blob/unified-routes/ga/docs/docs/file-system-route-api.md

---

@t2ca 

> Is it possible to insert props into pageContext like we did in gatsby-node with createPage?

Not at the moment. I've noted that request (I want it myself) for after the GA release and will probably rank it higher than other things as I also find it useful. The doc I linked above explicitly mentions that the File System Route API won't cover all use cases (now and probably ever).

**Comment by LekoArts** [comment]
In https://github.com/gatsbyjs/gatsby/pull/27728 I've added the ability to use prefixes and postfixes in paths, e.g. `/foo/prefix-{Model.bar}.js` or `/foo/{Model.bar}-postfix.js`

**Comment by czabaj** [comment]
@LekoArts I've finally tried dynamic categories but I encountered an error. I don't know whether I made something wrong, or I misunderstand how it works.

I have the following data source

```
query MyQuery {
  faqCategory {
    internal {
      type
    }
    questions {
      slug
    }
    slug
  }
}
```

which yields something like this

```
{
  "data": {
    "faqCategory": {
      "internal": {
        "type": "FaqCategory"
      },
      "questions": [
        {
          "slug": "question-slug-A"
        },
        {
          "slug": "question-slug-B"
        },
        {
          "slug": "question-slug-C"
        },
      ],
      "slug": "category-slug"
    }
  },
  "extensions": {}
}
```

Now, I create folders

- `src/pages/faq/{FaqCategory.slug}/index.js`
- `src/pages/faq/{FaqCategory.slug}/{FaqCategory.questions__slug}.js`

And I assume I will get

- page `fag/categorySlug` for all categories
- page `fag/categorySlug/categoryQuestionSlug` for _all questions inside categories_

In other words, it should work like

```
forEach category in categories
    create page category.slug
    forEach question in category.questions
        create page category.slug/question.slug
```

The category pages work, but the _nested collection_ fails with

```
Could not find value in the following node for key {FaqCategory.slug}/{FaqCategory.questions__slug} (transformed to slug/FaqCategory.questions.slug)
{
    "slug": "category-slug",
    "questions": [
        {
            "slug": "question-slug-A"
        },
        {
            "slug": "question-slug-B"
        },
        {
            "slug": "question-slug-C"
        },
}
```

Well, does the new FS API handle _collections_ inside dynamic directories?

**Comment by LekoArts** [comment]
@czabaj The changes were only available in the PR, so nothing wrong on your end. I published a new canary with the latest changes, you can install it with `npm install gatsby@unifiedroutes` :)

**Comment by herecydev** [comment]
@LekoArts Got an interesting problem here. Does the file system routing support optional pages?

Example: All cities have a name, they might optionally have a library

```
src/pages/{City.name}/index.js                                           // creates London, Paris, Tokyo
src/pages/{City.name}/{City.library__name}.js                     // fails when the city doesn't have a library
```

Error stems from [here](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-page-creator/src/create-pages-from-collection-builder.ts#L108). Would seem to me it could just safely ignore paths where data was not supplied.

What are your thoughts?

**Comment by LekoArts** [comment]
@herecydev Ah, interesting 🤔 I think for the majority it's the safer option to assume that this is an error, not something on purpose. I rather have explicit things rather than implicit guessing of "we could ignore this as we assume this is on purpose" 😅 

I rather have that fixed via the user explicitly setting a default through the schema customization API. Then `null` would be returned and we could not check against a falsy value but e.g. against `null` and let it pass.

Have to think about it however. What happens when you explicitly type that via schema API by the way?

**Comment by herecydev** [comment]
> What happens when you explicitly type that via schema API by the way?

Providing a value of null results in the path actually being "null", in addition the page is still attempted to be created, i.e. the props will still come through with the null value.

In the `gatsby-node` world, we would simply wrap a check before creating the templated page:

```
if(city.library) {
//then create the page...
}
```

> I rather have explicit things rather than implicit guessing of "we could ignore this as we assume this is on purpose"

What I would say here is that if you mispell the query `City.libraryyyy` it calls it out and says that the query doesn't resolve to anything. So it seems there is a safety net already there.

I would suggest using something like "{City.library}?" to denote that you anticipate it being optional but that's not possible in windows. Have a think and let me know, happy to be the guinea pig here (and more than happy to implement whatever is suggested)

**Comment by josephmarkus** [comment]
Has anyone got any timing metrics on this? Curious, if this way of creating pages is faster?

**Comment by talohana** [comment]
I might have missed something, but the [Migration: Programmatically created pages](https://www.gatsbyjs.com/blog/fs-route-api)  contains sorting by `frontmatter___date`, is this possible with the new API? because I can only see this is being removed from the example

**Comment by LekoArts** [comment]
> In the gatsby-node world, we would simply wrap a check before creating the templated page:

We could check against `null` specifically (so not falsy) and then skip the creation I guess. As `null` would mean you defined it.

> Has anyone got any timing metrics on this? Curious, if this way of creating pages is faster?

It should be similarly fast. Crunching out performance improvements is most probably possible.

> I might have missed something, but the Migration: Programmatically created pages contains sorting by frontmatter___date, is this possible with the new API? because I can only see this is being removed from the example

For the page creation the sorting is irrelevant. It's in there currently because the blog starter creates prev/next links so the order needs to be the same. As we currently don't support `pageContext` additions I'd say this use case isn't needed right now.

**Comment by Vacilando** [comment]
Will it ever be possible to insert props into pageContext like we did in gatsby-node with createPage (as asked in https://github.com/gatsbyjs/gatsby/issues/26375#issuecomment-718982239 above)? 

Or will those use cases have to continue using createPages in gatsby-node.js? 

Is it useful to open a ticket for that - or is there one already?
